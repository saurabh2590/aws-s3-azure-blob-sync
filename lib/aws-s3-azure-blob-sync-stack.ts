import * as cdk from 'aws-cdk-lib';
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as s3 from "aws-cdk-lib/aws-s3";
import * as logs from "aws-cdk-lib/aws-logs";
import {Construct} from 'constructs';
import * as path from "path";
import {LogGroupClass, RetentionDays} from "aws-cdk-lib/aws-logs";
import {S3EventSourceV2} from "aws-cdk-lib/aws-lambda-event-sources";
import {RemovalPolicy} from "aws-cdk-lib";

interface AwsS3AzureBlobSyncStackProps extends cdk.StackProps {
  bucket: string;
  azureStorageAccountName: string;
  azureStorageAccountKey: string;
  azureContainerName: string;
}

export class AwsS3AzureBlobSyncStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: AwsS3AzureBlobSyncStackProps) {
    super(scope, id, props);
    const {bucket, azureContainerName, azureStorageAccountName, azureStorageAccountKey} = props
    /**
     * Layers to publish Node Modules functions
     */
    const nodeModulesLayer = new lambda.LayerVersion(
      this,
      "node-modules-layer",
      {
        code: lambda.Code.fromAsset(
          path.join(__dirname, "../layers/node")
        ),
        compatibleRuntimes: [lambda.Runtime.NODEJS_20_X],
        license: "Apache-2.0",
        description: "Node Modules layer for the Lambda",
      }
    );

    const awsAzureSyncLambdaLogGroup = new logs.LogGroup(
      this,
      "awsAzureSyncLambdaLogGroup",
      {
        logGroupName: `${bucket}-awsAzureSyncLambda`,
        retention: RetentionDays.ONE_WEEK,
        logGroupClass: LogGroupClass.INFREQUENT_ACCESS,
        removalPolicy: RemovalPolicy.DESTROY,
      },
    );

    const awsAzureSyncLambda = new lambda.Function(this, "aws-azure-sync-lambda", {
      runtime: lambda.Runtime.NODEJS_20_X,
      code: lambda.Code.fromAsset("build"),
      timeout: cdk.Duration.seconds(60),
      logGroup: awsAzureSyncLambdaLogGroup,
      layers: [nodeModulesLayer],
      memorySize: 128,
      environment: {
        AZURE_STORAGE_ACCOUNT_NAME: azureStorageAccountName,
        AZURE_STORAGE_ACCOUNT_KEY: azureStorageAccountKey,
        AZURE_CONTAINER_NAME: azureContainerName
      },
      handler: "awsToAzureSync.handler",
      functionName: `s3-${bucket}-azure-${azureContainerName}-sync`
    });

    const s3Bucket = s3.Bucket.fromBucketName(this, `${bucket}-source`, bucket);
    s3Bucket.grantRead(awsAzureSyncLambda);

    awsAzureSyncLambda.addEventSource(
      new S3EventSourceV2(s3Bucket, {
        events: [s3.EventType.OBJECT_CREATED]
      })
    )
  }
}
