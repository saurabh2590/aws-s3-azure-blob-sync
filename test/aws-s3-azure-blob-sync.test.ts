import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import {AwsS3AzureBlobSyncStack} from "../lib/aws-s3-azure-blob-sync-stack";
import "jest-cdk-snapshot";
// example test. To run these tests, uncomment this file along with the
// example resource in lib/aws-s3-azure-blob-sync-stack.ts
test('SQS Queue Created', () => {
  const app = new cdk.App();
    // WHEN
  const stack = new AwsS3AzureBlobSyncStack(app, 'MyTestStack', {
    bucket: "some-bucket",
    azureStorageAccountName: "azure-storage-account-name",
    azureStorageAccountKey: "azure-storage-account-key",
    azureContainerName: "azure-container-name"
  });
    // THEN
  const template = Template.fromStack(stack);

  expect(stack).toMatchCdkSnapshot({
    ignoreAssets: true,
  });
});
