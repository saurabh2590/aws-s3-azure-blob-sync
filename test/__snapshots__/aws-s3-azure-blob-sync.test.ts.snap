// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SQS Queue Created 1`] = `
{
  "Resources": {
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db8347ECC3691": {
      "DependsOn": [
        "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36",
        "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
      ],
      "Properties": {
        "Code": Any<Object>,
        "Description": "AWS CloudFormation handler for "Custom::S3BucketNotifications" resources (@aws-cdk/aws-s3)",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
            "Arn",
          ],
        },
        "Runtime": "python3.11",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:PutBucketNotification",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "s3:GetBucketNotification",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36",
        "Roles": [
          {
            "Ref": "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "awsAzureSyncLambdaLogGroupDC3DA102": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "LogGroupClass": "INFREQUENT_ACCESS",
        "LogGroupName": "some-bucket-awsAzureSyncLambda",
        "RetentionInDays": 7,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "awsazuresynclambda91954E56": {
      "DependsOn": [
        "awsazuresynclambdaServiceRoleDefaultPolicy0FAC98E7",
        "awsazuresynclambdaServiceRoleCDFABF5D",
      ],
      "Properties": {
        "Code": Any<Object>,
        "Environment": {
          "Variables": {
            "AZURE_CONTAINER_NAME": "azure-container-name",
            "AZURE_STORAGE_ACCOUNT_KEY": "azure-storage-account-key",
            "AZURE_STORAGE_ACCOUNT_NAME": "azure-storage-account-name",
          },
        },
        "FunctionName": "s3-some-bucket-azure-azure-container-name-sync",
        "Handler": "awsToAzureSync.handler",
        "Layers": [
          {
            "Ref": "nodemoduleslayer07C837B6",
          },
        ],
        "LoggingConfig": {
          "LogGroup": {
            "Ref": "awsAzureSyncLambdaLogGroupDC3DA102",
          },
        },
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "awsazuresynclambdaServiceRoleCDFABF5D",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "awsazuresynclambdaServiceRoleCDFABF5D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "awsazuresynclambdaServiceRoleDefaultPolicy0FAC98E7": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::some-bucket",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::some-bucket/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "awsazuresynclambdaServiceRoleDefaultPolicy0FAC98E7",
        "Roles": [
          {
            "Ref": "awsazuresynclambdaServiceRoleCDFABF5D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "nodemoduleslayer07C837B6": {
      "Properties": {
        "CompatibleRuntimes": [
          "nodejs20.x",
        ],
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "792a3d33c06808b549ea93234a5680aa376222e85006e30ae3f6835647e7f9e6.zip",
        },
        "Description": "Node Modules layer for the Lambda",
        "LicenseInfo": "Apache-2.0",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "somebucketsourceAllowBucketNotificationsToMyTestStackawsazuresynclambda25F128192EB95386": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "awsazuresynclambda91954E56",
            "Arn",
          ],
        },
        "Principal": "s3.amazonaws.com",
        "SourceAccount": {
          "Ref": "AWS::AccountId",
        },
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":s3:::some-bucket",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "somebucketsourceNotifications3FA98829": {
      "DependsOn": [
        "somebucketsourceAllowBucketNotificationsToMyTestStackawsazuresynclambda25F128192EB95386",
      ],
      "Properties": {
        "BucketName": "some-bucket",
        "Managed": false,
        "NotificationConfiguration": {
          "LambdaFunctionConfigurations": [
            {
              "Events": [
                "s3:ObjectCreated:*",
              ],
              "LambdaFunctionArn": {
                "Fn::GetAtt": [
                  "awsazuresynclambda91954E56",
                  "Arn",
                ],
              },
            },
          ],
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "BucketNotificationsHandler050a0587b7544547bf325f094a3db8347ECC3691",
            "Arn",
          ],
        },
        "SkipDestinationValidation": false,
      },
      "Type": "Custom::S3BucketNotifications",
    },
  },
}
`;
